pkg/operator/controller/dns_status.go

```diff
@@ -0,0 +1,146 @@
+		progressingCondition.Status = operatorv1.ConditionFalse
+		progressingCondition.Reason = "DaemonSetAvailable"
+		progressingCondition.Message = "All expected nodes running DaemonSet pod"
+	} else {
+		progressingCondition.Status = operatorv1.ConditionTrue
+		progressingCondition.Reason = "Reconciling"
+		progressingCondition.Message = fmt.Sprintf("%d hosts running a DaemonSet pod, want %d",
+			ds.Status.NumberAvailable, ds.Status.DesiredNumberScheduled)
+	}
+
+	return setDNSStatusCondition(conditions, progressingCondition)
+}
+
+// computeDNSAvailableCondition computes the dns Available status condition
+// based on the status of ds.
+func computeDNSAvailableCondition(conditions []operatorv1.OperatorCondition, ds *appsv1.DaemonSet) []operatorv1.OperatorCondition {
+	availableCondition := &operatorv1.OperatorCondition{
+		Type:   operatorv1.OperatorStatusTypeAvailable,
+		Status: operatorv1.ConditionUnknown,
+	}
+	if ds.Status.NumberAvailable > 0 {
+		availableCondition.Status = operatorv1.ConditionTrue
+		availableCondition.Reason = "MinimumDaemonSetPodsAvailable"
+		availableCondition.Message = "Minimum number of nodes running DaemonSet pod"
+	} else {
+		availableCondition.Status = operatorv1.ConditionFalse
+		availableCondition.Reason = "DaemonSetUnavailable"
+		availableCondition.Message = "DaemonSet pod not running on any nodes"
+	}
+
+	return setDNSStatusCondition(conditions, availableCondition)
+}
+
+// setDNSStatusCondition returns the dns condition result
+// of setting the specified condition in the given slice of conditions.
+func setDNSStatusCondition(oldConditions []operatorv1.OperatorCondition, condition *operatorv1.OperatorCondition) []operatorv1.OperatorCondition {
+	condition.LastTransitionTime = metav1.Now()
+
+	var newConditions []operatorv1.OperatorCondition
+
+	found := false
+	for _, c := range oldConditions {
+		if condition.Type == c.Type {
+			if condition.Status == c.Status &&
+				condition.Reason == c.Reason &&
+				condition.Message == c.Message {
+				// The dns status condition has not changed.
+				return oldConditions
+			}
+
+			found = true
+			newConditions = append(newConditions, *condition)
+		} else {
+			newConditions = append(newConditions, c)
+		}
+	}
+	if !found {
+		newConditions = append(newConditions, *condition)
+	}
+
+	return newConditions
+}
+
+// dnsStatusesEqual compares two DNSStatus values.  Returns true
+// if the provided values should be considered equal for the purpose of determining
+// whether an update is necessary, false otherwise.
+func dnsStatusesEqual(a, b operatorv1.DNSStatus) bool {
+	conditionCmpOpts := []cmp.Option{
+		cmpopts.IgnoreFields(operatorv1.OperatorCondition{}, "LastTransitionTime"),
+		cmpopts.EquateEmpty(),
+		cmpopts.SortSlices(func(a, b operatorv1.OperatorCondition) bool { return a.Type < b.Type }),
+	}
+	if !cmp.Equal(a.Conditions, b.Conditions, conditionCmpOpts...) || a.ClusterIP != b.ClusterIP || a.ClusterDomain != b.ClusterDomain {
+		return false
+	}
```

Comment:
In `statusesEqual`, we do the following:
```go
	if foo is different {
		return false
	}
	if bar is different {
		return false
	}
	// ...
	return true
```
I find that approach easier to read.  I don't object if you prefer a single `if foo is different || if bar different || ... {` statement though.  However, if you do go in that direction, you could simplify further by doing away with the `if` and simply returning a Boolean expression:
```go
	return foo is the same && bar is the same && ...
```
Concretely:
```go
	return cmp.Equal(a.Conditions, b.Conditions, conditionCmpOpts...) && a.ClusterIP == b.ClusterIP && a.ClusterDomain == b.ClusterDomain
```

Reply:
I agree with your recommendation and find the code easier to read that way. I'll update.